sql/src/main/java/io/crate/breaker/RamAccountingContext.java

```diff
@@ -146,21 +147,36 @@ public long totalBytes() {
 
     /**
      * Close the context and adjust the breaker.
-     * A remaining flush buffer will not be flushed to avoid breaking on close.
      * (all ram operations expected to be finished at this point)
      */
     public void close() {
         if (closed) {
             return;
         }
         closed = true;
+        release();
+    }
+
+    /**
+     * Release all the bytes that have been flushed to the breaker so far, and the bytes that are in the buffer "to be
+     * flushed" are not accounted for in the breaker anymore, but only in the #{@link RamAccountingContext#totalBytes}.
+     *
+     * The purpose of this method is to substract everything that this context added to the breaker so far in order
+     * to be reused in a multi-phase operation where a subsequent phase needs to make decisions based on the available
+     * memory after the previous phase completed (and needs to be unloaded/released from the breaker)
+     */
+    public void release() {
         if (totalBytes.get() != 0) {
             if (logger.isTraceEnabled() && totalBytes() > FLUSH_BUFFER_SIZE) {
                 logger.trace("context: {} bytes; breaker: {} of {} bytes", totalBytes(), breaker.getUsed(), breaker.getLimit());
             }
-            breaker.addWithoutBreaking(-totalBytes.get());
+            long bytesToRelease = totalBytes.get() - releasedBytes.get();
+            breaker.addWithoutBreaking(-bytesToRelease);
+            releasedBytes.addAndGet(bytesToRelease);
         }
-        totalBytes.addAndGet(flushBuffer.getAndSet(0));
+        long bytesInFlushBuffer = flushBuffer.getAndSet(0);
+        totalBytes.addAndGet(bytesInFlushBuffer);
+        releasedBytes.addAndGet(bytesInFlushBuffer);
```

Comment:
why do we need releasedBytes? also I think this is wrong as `long bytesToRelease = totalBytes.get() - releasedBytes.get();` would account the bytes not-yet-flushed 2 times..

Reply:
We need the releasedBytes for the case where we call `release` multiple times (in which case on 2nd, 3rd etc calls we will only need to release what we accumulated since the previous call to `release`) 
I don't think we account for not flushed bytes twice because we empty the flush buffer when we release and we account for it in the releasedBytes _separately_ than the totalBytes

Reply:
hm ok but I still don't see the need for the releasedBytes. `totalBytes()` is used only for statistics to see how much memory an operations has used. I think we should substract the totalBytes on release, otherwise the statistics would be wrong. maybe this leads to slightly different implementations of `close()` and `release()`